src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,88 @@
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    internal abstract class AbstractConvertLinqQueryToForEachProvider<TQueryExpression> : CodeRefactoringProvider
+        where TQueryExpression : SyntaxNode
+    {
+        protected abstract string Title { get; }
+
+        protected abstract bool TryConvert(
+            TQueryExpression queryExpression,
+            SemanticModel semanticModel,
+            ISemanticFactsService semanticFacts,
+            CancellationToken cancellationToken,
+            out DocumentUpdateInfo documentUpdate);
+
+        protected abstract TQueryExpression FindNodeToRefactor(SyntaxNode root, TextSpan span);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var queryExpression = FindNodeToRefactor(root, context.Span);
+            if (queryExpression == null || queryExpression.ContainsDiagnostics)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var semanticFacts = document.GetLanguageService<ISemanticFactsService>();
+            if (TryConvert(queryExpression, semanticModel, semanticFacts, cancellationToken, out DocumentUpdateInfo documentUpdateInfo))
+            {
+                context.RegisterRefactoring(
+                    new MyCodeAction(
+                        Title,
+                        c => Task.FromResult(document.WithSyntaxRoot(documentUpdateInfo.UpdateRoot(root)))));
+            }
+        }
+
+        /// <summary>
+        /// Handles information about updating the document with the refactoring.
+        /// </summary>
+        internal sealed class DocumentUpdateInfo
+        {
+            private readonly SyntaxNode _source;
+            private readonly ImmutableArray<SyntaxNode> _destinations;
+
+            public DocumentUpdateInfo(SyntaxNode source, SyntaxNode destination) : this(source, new[] { destination })
+            {
+            }
+
+            public DocumentUpdateInfo(SyntaxNode source, IEnumerable<SyntaxNode> destinations)
+            {
+                _source = source;
+                _destinations = ImmutableArray.CreateRange(destinations);
+            }
+
+            /// <summary>
+            /// Updates the root of the docuemtn with the document update.
+            /// </summary>
+            public SyntaxNode UpdateRoot(SyntaxNode root)
+            {
+                return root.ReplaceNode(_source, _destinations);
+            }
+        }
+
+        protected sealed class MyCodeAction : CodeAction.DocumentChangeAction
```

Comment:
>My [](start = 31, length = 2)

It feels strange to see "My" prefix on a class name. Is this a well established pattern? Consider removing otherwise.

Reply:
Yes, it is widespread. Using just 'CodeAction' conflicts with the basic CodeAction class. So, the pattern is 'MyCodeAction'

---
In reply to: [176928161](https://github.com/dotnet/roslyn/pull/25362#discussion_r176928161) [](ancestors = 176928161)